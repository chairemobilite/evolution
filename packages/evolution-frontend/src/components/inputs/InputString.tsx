/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';

import { _isBlank } from 'chaire-lib-common/lib/utils/LodashExtensions';
import { InputStringType } from '../../services/widgets';
import { parseString } from '../../utils/helpers';
import { UserInterviewAttributes } from 'evolution-common/lib/services/interviews/interview';
import { FrontendUser } from 'chaire-lib-frontend/lib/services/auth/user';

export interface InputStringProps<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> {
    id: string;
    onValueChange?: (e: any) => void;
    value?: string;
    interview: UserInterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
    // TODO There's also a path in widgetConfig, but this one comes from the props of the question. See if it's always the same and use the one from widgetConfig if necessary
    path: string;
    user: FrontendUser;
    inputRef?: React.LegacyRef<HTMLInputElement>;
    size?: 'small' | 'medium' | 'large';
    widgetConfig: InputStringType;
    updateKey: number;
}

const getStateValue = <CustomSurvey, CustomHousehold, CustomHome, CustomPerson>(
    props: InputStringProps<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>
) => {
    const value = _isBlank(props.value)
        ? parseString(props.widgetConfig.defaultValue, props.interview, props.path, props.user)
        : props.value;
    return _isBlank(value) ? '' : value;
};

export const InputString = <CustomSurvey, CustomHousehold, CustomHome, CustomPerson>(
    props: InputStringProps<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>
) => {
    const [value, setValue] = React.useState(getStateValue(props));
    // Update the value with props when updateKey prop is updated. Not using the
    // `key` to avoid loosing the identity of the component (which causes the
    // focus to get lost)
    React.useEffect(() => {
        setValue(getStateValue(props));
    }, [props.updateKey]);
    return (
        <input
            autoComplete="none"
            type="text"
            className={`apptr__form-input apptr__input-string input-${props.size || 'large'}`}
            name={props.id}
            id={props.id}
            value={value}
            onBlur={props.onValueChange}
            onChange={(e) => setValue(e.target.value)}
            ref={props.inputRef}
            maxLength={props.widgetConfig.maxLength ? props.widgetConfig.maxLength : 255}
        />
    );
};

export default InputString;
