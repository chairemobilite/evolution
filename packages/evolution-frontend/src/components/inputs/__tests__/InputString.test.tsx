/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import TestRenderer from 'react-test-renderer';
import { mount } from 'enzyme';

import { InputString } from '../InputString';
import { interviewAttributes } from './interviewData.test';

const userAttributes = {
    id: 1,
    username: 'foo',
    preferences: {  },
    serializedPermissions: [],
    isAuthorized: () => true,
    is_admin: false,
    pages: [],
    showUserInfo: true
}

test('Should correctly render InputString with all parameters', () =>{

    const widgetConfig = {
        type: 'question' as const,
        twoColumns: true,
        path: 'test.foo',
        inputType: 'string' as const,
        datatype: 'string' as const,
        maxLength: 20,
        defaultValue: 'test',
        containsHtml: true,
        label: {
            fr: `Texte en français`,
            en: `English text`
        }
    }

    const wrapper = TestRenderer.create(
        <InputString
            id={'test'}
            onValueChange={() => { /* nothing to do */}}
            widgetConfig={widgetConfig}
            value={undefined}
            inputRef={React.createRef()}
            size='medium'
            updateKey={0}
            interview={interviewAttributes}
            path={'path'}
            user={userAttributes}
        />
    );
    expect(wrapper).toMatchSnapshot();
});

test('Should correctly render InputString with default value as function', () =>{

    const widgetConfig = {
        type: 'question' as const,
        twoColumns: true,
        path: 'test.foo',
        inputType: 'string' as const,
        datatype: 'string' as const,
        maxLength: 20,
        defaultValue: jest.fn().mockReturnValue('fctDefault'),
        containsHtml: true,
        label: {
            fr: `Texte en français`,
            en: `English text`
        }
    }

    const wrapper = TestRenderer.create(
        <InputString
            id={'test'}
            onValueChange={() => { /* nothing to do */}}
            widgetConfig={widgetConfig}
            value={undefined}
            inputRef={React.createRef()}
            size='medium'
            updateKey={0}
            interview={interviewAttributes}
            path={'path'}
            user={userAttributes}
        />
    );
    expect(wrapper).toMatchSnapshot();
    expect(widgetConfig.defaultValue).toHaveBeenCalledWith(interviewAttributes, 'path', userAttributes);
});

test('Should correctly render InputString with base parameters', () =>{

    const widgetConfig = {
        type: 'question' as const,
        path: 'test.foo',
        inputType: 'string' as const,
        label: {
            fr: `Texte en français`,
            en: `English text`
        }
    }
    const wrapper = TestRenderer.create(
        <InputString
            id={'test'}
            widgetConfig={widgetConfig}
            value='value'
            updateKey={0}
            interview={interviewAttributes}
            path={'path'}
            user={userAttributes}
        />
    );
    expect(wrapper).toMatchSnapshot();
});

test('Should correctly render InputString with a value of 0', () =>{

    const widgetConfig = {
        type: 'question' as const,
        path: 'test.foo',
        inputType: 'string' as const,
        label: {
            fr: `Texte en français`,
            en: `English text`
        }
    }
    const wrapper = TestRenderer.create(
        <InputString
            id={'test'}
            widgetConfig={widgetConfig}
            value={0 as any}
            updateKey={0}
            interview={interviewAttributes}
            path={'path'}
            user={userAttributes}
        />
    );
    expect(wrapper).toMatchSnapshot();
});

test('Test update value through props', () => {
    const widgetConfig = {
        type: 'question' as const,
        path: 'test.foo',
        inputType: 'string' as const,
        label: {
            fr: `Texte en français`,
            en: `English text`
        }
    }
    const testId = 'test';
    const stringInput = mount(<InputString
        id={testId}
        widgetConfig={widgetConfig}
        value='value'
        updateKey={0}
        interview={interviewAttributes}
        path={'path'}
        user={userAttributes}
    />);

    // Validate initial values
    const inputElement = stringInput.find({id: `${testId}`, type: 'text'});
    expect(inputElement).toBeTruthy();
    expect(inputElement.getDOMNode<HTMLInputElement>().value).toBe('value');

    // Change value 'manually'
    const newValue = 'newVal';
    inputElement.getDOMNode<HTMLInputElement>().value = newValue;
    inputElement.simulate('change');
    expect(inputElement.getDOMNode<HTMLInputElement>().value).toBe(newValue);

    // Change value through props without changing updateKey, should not change
    stringInput.setProps({
        id: testId,
        widgetConfig,
        value: 'updateByClient',
        updateKey: 0,
        interview: interviewAttributes,
        path: 'path',
        user: userAttributes
    });
    expect(inputElement.getDOMNode<HTMLInputElement>().value).toBe(newValue);

    // Change value through props and change updateKey, should be updated
    const updateByServerVal = 'updateByServer';
    stringInput.setProps({
        id: testId,
        widgetConfig,
        value: updateByServerVal,
        updateKey: 1
    });
    expect(inputElement.getDOMNode<HTMLInputElement>().value).toBe(updateByServerVal);
});