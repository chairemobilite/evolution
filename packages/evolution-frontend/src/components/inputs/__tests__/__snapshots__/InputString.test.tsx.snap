// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should correctly render InputString with a value of 0 1`] = `
<input
  autoComplete="none"
  className="apptr__form-input apptr__input-string input-large"
  id="test"
  maxLength={255}
  name="test"
  onBlur={[MockFunction]}
  onChange={[Function]}
  style={
    Object {
      "textTransform": "none",
    }
  }
  type="text"
  value={0}
/>
`;

exports[`Should correctly render InputString with all parameters 1`] = `
<input
  autoComplete="none"
  className="apptr__form-input apptr__input-string input-medium"
  id="test"
  maxLength={20}
  name="test"
  onBlur={[Function]}
  onChange={[Function]}
  style={
    Object {
      "textTransform": "lowercase",
    }
  }
  type="text"
  value="test"
/>
`;

exports[`Should correctly render InputString with base parameters 1`] = `
<input
  autoComplete="none"
  className="apptr__form-input apptr__input-string input-large"
  id="test"
  maxLength={255}
  name="test"
  onBlur={[MockFunction]}
  onChange={[Function]}
  style={
    Object {
      "textTransform": "none",
    }
  }
  type="text"
  value="value"
/>
`;

exports[`Should correctly render InputString with default value as function 1`] = `
<input
  autoComplete="none"
  className="apptr__form-input apptr__input-string input-medium"
  id="test"
  maxLength={20}
  name="test"
  onBlur={[Function]}
  onChange={[Function]}
  style={
    Object {
      "textTransform": "none",
    }
  }
  type="text"
  value="fctDefault"
/>
`;
